runtime/src/main/java/org/corfudb/runtime/view/stream/CheckpointStreamAddressSpace.java

```diff
@@ -0,0 +1,175 @@
+package org.corfudb.runtime.view.stream;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Function;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.corfudb.protocols.logprotocol.CheckpointEntry;
+import org.corfudb.protocols.wireprotocol.DataType;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.util.Utils;
+
+@Slf4j
+/**
+ * The Checkpoint Stream Address Space resolves the space of addresses of a checkpoint(ed) stream,
+ * therefore it is capable of following backpointers and resolving the checkpoint at the same time,
+ * i.e., determine how to traverse the cp stream
+ *
+ * Created by amartinezman on 4/18/18.
+ */
+public class CheckpointStreamAddressSpace extends StreamAddressSpace {
+
+    /** Info on checkpoint we used for initial stream replay,
+     *  other checkpoint-related info & stats.  Hodgepodge, clarify.
+     */
+    @Getter
+    public long lastCPAddress;
+    UUID checkpointSuccessId = null;
+    long checkpointSuccessStartAddr = Address.NEVER_READ;
+    long checkpointSuccessEndAddr = Address.NEVER_READ;
+    long checkpointSuccessNumEntries = 0L;
+    long checkpointSuccessBytes = 0L;
+
+    /** The address the current checkpoint snapshot was taken at.
+     *  The checkpoint guarantees for this stream there are no entries
+     *  between checkpointSuccessStartAddr and checkpointSnapshotAddress.
+     */
+    long checkpointSnapshotAddress = Address.NEVER_READ;
+
+    public CheckpointStreamAddressSpace(UUID id, CorfuRuntime runtime) {
+        super(id, runtime);
+        lastCPAddress = Address.NON_ADDRESS;
+    }
+
+    @Override
+    public long getLastAddressSynced() {
+        if (maxInd == -1) {
+            return Address.NON_ADDRESS;
+        } else {
+            return lastCPAddress;
+        }
+    }
+
+    private void setLastCheckpointedAddress(ILogData firstCPEntry) {
+        if (firstCPEntry.hasCheckpointMetadata()) {
+            CheckpointEntry cpEntry = (CheckpointEntry)
+                    firstCPEntry.getPayload(runtime);
+
+            if (cpEntry.getCpType() == CheckpointEntry.CheckpointEntryType.START) {
+                lastCPAddress = Long.decode(cpEntry.getDict().get(CheckpointEntry.CheckpointDictKey.START_LOG_ADDRESS));
+                return;
+            }
+        }
+    }
+
+    @Override
+    public long getCurrentPointer() {
+        // The current pointer refers to the current version of the object, which is in fact related
+        // to the last global address synced for this checkpoint stream.
+        return getLastAddressSynced();
+    }
+
+    @Override
+    public int findAddresses(long oldTail, long newTail, Function<Long, ILogData> readFn) {
```

Comment:
This seems to be somewhat redundant to the code in BackpointerStreamAddressSpace::findAddresses (?). Can we make sure that we reuse the code? (This BackpointerOp logic used to handle checkpoint/noncheckpoint streams).